src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -535,6 +536,7 @@ private async Task<bool> PerformAuditAsync(IEnumerable<RestoreTargetGraph> graph
         {
             telemetry.StartIntervalMeasure();
             var audit = new AuditUtility(
+                _request.Project.TargetFrameworks.SelectMany(t => t.DownloadDependencies),
```

Comment:
LINQ always causes heap allocations. Firstly in the delegate method it needs to pass, secondly the enumerator that it internally uses to iterate all the items in `TargetFrameworks`, and finally there will be another allocation when the result of the `SelectMany` is enumerated. The return type of SelectMany may even be a fourth allocation.  It is a fixed number of allocations per call, but it will happen for each project in the restore.

So, I'm thinking that passing through the `TargetFrameworks` collection, and then having audit iterate that (using `.NoAllocEnumerate()`, or whatever that API is called), and only read the `DownloadDependencies` property will allow us to check PackageDownloads without any allocations, and therefore fewer garbage collections when there are a huge number of projects.

Unfortunately, it will make setting up tests harder, so it's a perf trade-off. audit runs in restore, which is our most perf-critical path.  I hope other people in the team speak up if they think I'm being too nit-picky about 3 or 4 allocations.

Reply:
The "equivalent" way of doing the audit to PR is to actually audit the downloaded packages as well. 

In this case it's not as relevant, but if that was represented in these types, maybe we wouldn't have needed to pass anything new. 
Given that's a different refactoring that I might need to do as part of https://github.com/NuGet/Home/issues/7722, and this PR would carry more urgency than that, I think it's ok if that refactoring is not done here but something to consider.

Reply:
I agree with Andy that making a copy of these is not a good idea.  Passing along the collection of TargetFrameworks isn't ideal either, maybe we should just pass the entire `PackageSpec` or perhaps an array of the lists?

```csharp
var downloadDependencyLists = new ImmutableArray<DownloadDependency>[_request.Project.TargetFrameworks.Count];

for (int i = 0; i < _request.Project.TargetFrameworks.Count; i++)
{
  downloadDependencyLists[i] = _request.Project.TargetFrameworks[i].DownloadDependencies;
}
```

Reply:
Thanks all for the ideas. I think @nkolev92 is right, refactoring might have been a good idea. But since we want to merge this ASAP, maybe we could do it as part of https://github.com/NuGet/Home/issues/7722.